primitives/block/src/equivocation_proof.rs

```diff
@@ -0,0 +1,423 @@
+/// validator created two macro block proposals at the same height, in the same round.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, SerializeContent)]
+pub struct DoubleProposalProof {
+    header1: MacroHeader,
+    header2: MacroHeader,
+    justification1: SchnorrSignature,
+    justification2: SchnorrSignature,
+    prev_vrf_seed1: VrfSeed,
+    prev_vrf_seed2: VrfSeed,
+}
+
+impl DoubleProposalProof {
+    /// The maximum size of a double proposal proof.
+    pub const SIZE: usize =
+        2 * MacroHeader::MAX_SIZE + 2 * SchnorrSignature::SIZE + 2 * VrfSeed::SIZE;
+
+    pub fn new(
+        mut header1: MacroHeader,
+        mut justification1: SchnorrSignature,
+        mut prev_vrf_seed1: VrfSeed,
+        mut header2: MacroHeader,
+        mut justification2: SchnorrSignature,
+        mut prev_vrf_seed2: VrfSeed,
+    ) -> DoubleProposalProof {
+        let hash1: Blake2bHash = header1.hash();
+        let hash2: Blake2bHash = header2.hash();
+        if hash1 > hash2 {
+            mem::swap(&mut header1, &mut header2);
+            mem::swap(&mut justification1, &mut justification2);
+            mem::swap(&mut prev_vrf_seed1, &mut prev_vrf_seed2);
+        }
+        DoubleProposalProof {
+            header1,
+            header2,
+            justification1,
+            justification2,
+            prev_vrf_seed1,
+            prev_vrf_seed2,
+        }
+    }
+
+    pub fn block_number(&self) -> u32 {
+        self.header1.block_number
+    }
+    pub fn round(&self) -> u32 {
+        self.header1.round
+    }
+    pub fn header1_hash(&self) -> Blake2bHash {
+        self.header1.hash()
+    }
+    pub fn header2_hash(&self) -> Blake2bHash {
+        self.header2.hash()
+    }
+    pub fn prev_vrf_seed1(&self) -> &VrfSeed {
+        &self.prev_vrf_seed1
+    }
+    pub fn prev_vrf_seed2(&self) -> &VrfSeed {
+        &self.prev_vrf_seed2
+    }
+
+    /// Verify the validity of a double proposal proof.
+    pub fn verify(&self, signing_key: &SchnorrPublicKey) -> Result<(), EquivocationProofError> {
+        let hash1: Blake2bHash = self.header1.hash();
+        let hash2: Blake2bHash = self.header2.hash();
+
+        // Check that the headers are not equal and in the right order:
+        match hash1.cmp(&hash2) {
+            Ordering::Less => {}
+            Ordering::Equal => return Err(EquivocationProofError::SameHeader),
+            Ordering::Greater => return Err(EquivocationProofError::WrongOrder),
+        }
+
+        if self.header1.block_number != self.header2.block_number
+            || self.header1.round != self.header2.round
+            || self.header1.seed.entropy() != self.header2.seed.entropy()
+        {
+            return Err(EquivocationProofError::SlotMismatch);
+        }
+
+        if let Err(e) = self.header1.seed.verify(&self.prev_vrf_seed1, signing_key) {
+            error!("DoubleProposalProof: VrfSeed 1 failed to verify: {:?}", e);
```

Comment:
Use `tracing style`:
```suggestion
        if let Err(error) = self.header1.seed.verify(&self.prev_vrf_seed1, signing_key) {
            error!(?error, "DoubleProposalProof: VrfSeed 1 failed to verify");
```